### ISSUE TITLE
Bad parenthesis in UNION under HSQLDB

### ISSUE THREAD

**Issue by knutwannheden** [jOOQ/jOOQ#9187]
A recursive CTE query like the following:

```java
Field<Integer> a = field(name("t", "a"), Integer.class);
Field<String> b = field(name("t", "b"), String.class);

CommonTableExpression<Record2<Integer, String>> t =
name("t").fields("a", "b").as(
    select(
        inline(1),
        inline("a").cast(VARCHAR(15))
    )
    .unionAll(
        select(
            a.add(inline(1)),
            b.concat(inline("a")).cast(VARCHAR(15))
        )
        .from(table(name("t")))
        .where(a.lt(inline(10)))
    )
);

Result<Record> result=
create().withRecursive(t)
        .select()
        .from(t)
        .fetch();

```

is not rendered correctly for HSQLDB:

```sql
with recursive "t"("a", "b") as (
  (
    select 
      1, 
      cast('a' as varchar(15))
    from (values(1))
  )
  union all (
    select 
      ("t"."a" + 1), 
      cast(("t"."b" || 'a') as varchar(15))
    from "t"
    where "t"."a" < 10
  )
)
select 
  "t"."a", 
  "t"."b"
from "t"
```

The problem is the parenthesis pair around the `SELECT`s before *and after* the `UNION ALL`. The extra parenthesis in either of these places causes a `SQL Error [42501]: user lacks privilege or object not found: t`.

With both parenthesis pairs removed the query parses, but runs into an OutOfMemoryError because HSQLDB requires a `UNION` here (see also https://sourceforge.net/p/hsqldb/bugs/1467/).

**Comment by knutwannheden** [comment]
For a separate issue: In HSQLDB
```sql
 select 
      1, 
      cast('a' as varchar(15))
    from (values(1))
```
can be abbreviated as
```sql
values(1, cast('a' as varchar(15)))
```

**Comment by lukaseder** [comment]
The actual issue here is probably also a bug in HSQLDB, but we should be able to better control the output of our unions and avoid parentheses where they aren't strictly required.

I'm reluctant to fix this in the current implementation. I would rather wait for our new query object model, which should implicitly fix everything that went wrong in the existing `SelectQueryImpl` implementation, as far as unions are concerned.

The existing implementation inherits my original misunderstanding of how unions and other set operations work in SQL, and what exactly is a union subquery. My original understanding matched Oracle's, but other dialects disagree with that.

Related issues (just some examples) include:

- https://github.com/jOOQ/jOOQ/issues/3791
- https://github.com/jOOQ/jOOQ/issues/4856

There are many more, which I'm hoping to be able to fix with the new query object model.

> can be abbreviated as

I wouldn't do that. The user writes `select(a, b)` and they would expect jOOQ to generate exactly that, and not be clever about some alternative syntax, which risks not being exactly the same thing in some edge cases.

We already have a feature request to support the `values()` table constructor wherever `Select` is possible: https://github.com/jOOQ/jOOQ/issues/5871

The jOOQ philosophy is:

- Always generate *exactly* what the user specifies through the DSL
- If that is not supported in the target dialect, *only then* emulate it through something else

**Comment by knutwannheden** [comment]
> > can be abbreviated as
> 
> I wouldn't do that. The user writes `select(a, b)` and they would expect jOOQ to generate exactly that, and not be clever about some alternative syntax, which risks not being exactly the same thing in some edge cases.
> 
> We already have a feature request to support the `values()` table constructor wherever `Select` is possible: #5871
> 
> The jOOQ philosophy is:
> 
> * Always generate _exactly_ what the user specifies through the DSL
> * If that is not supported in the target dialect, _only then_ emulate it through something else

Yes, of course. For some reason I was thinking the original query was using `VALUES`...

**Comment by lukaseder** [comment]
Could be that this is fixed via #3676

**Comment by lukaseder** [comment]
Seems so. Now, our integration test `testRecursiveCTEMultiple` doesn't fail with syntax errors. It fails because the recursion never stops on this query :)

```sql
with recursive 
  "t1"("f1", "f2") as (
    select 
      1, 
      cast('a' as varchar(15))
    from (select 1 as dual from information_schema.system_users limit 1) as dual
    union all select 
      ("t1"."f1" + 1), 
      cast(("t1"."f2" || 'a') as varchar(15))
    from "t1"
    where "t1"."f1" < 10
  ), 
  "t2"("g1", "g2") as (
    select 
      1, 
      cast('b' as varchar(15))
    from (select 1 as dual from information_schema.system_users limit 1) as dual
    union all select 
      ("t2"."g1" + 1), 
      cast(("t2"."g2" || 'b') as varchar(15))
    from "t2"
    where "t2"."g1" < 10
  )
select 
  "t1"."f1", 
  "t1"."f2", 
  "t2"."g1", 
  "t2"."g2"
from "t1"
  join "t2"
    on "t1"."f1" = "t2"."g1"
```

Perhaps an HSQLDB bug. Will investigate further

**Comment by knutwannheden** [comment]
See this comment above.

> With both parenthesis pairs removed the query parses, but runs into an OutOfMemoryError because HSQLDB requires a `UNION` here (see also https://sourceforge.net/p/hsqldb/bugs/1467/).

**Comment by lukaseder** [comment]
Ah yes indeed

**Comment by lukaseder** [comment]
Well, I have already created the new issue :) https://sourceforge.net/p/hsqldb/bugs/1565/

**Comment by lukaseder** [comment]
Anyway, it's documented here, and users can use a `UNION` explicitly to work around HSQLDB issues 1467 / 1565. Starting with 3.13.0, this issue no longer persists, as we're no longer generating the extra parentheses out of the box, so I'm closing this as it is fixed with #3676
